serles/challenge.py

```diff
@@ -127,6 +131,46 @@ def http_challenge(challenge):  # RFC8555 §8.3
 
     return None, None  # no error occurred :)
 
+def dns_challenge(challenge):  # RFC8555 §8.4
+    """ verify a DNS-01 Challenge
+
+    Args:
+        challenge (Challenge): The DNS-01 challenge to verify.
+
+    Returns:
+        tuple(str,str): problem detail type of the error and  textual
+        description, or (None,None).
+    """
+    host = challenge.authorization.identifier.value
+
+    # When it´s a wildcard domain, remove the wildcard
+    if host.startswith("*."):
+        host = host[2:]
+
+    # Try to resolve the _acme-challenge record
+    try:
+        answers = dns.resolver.query(f"_acme-challenge.{host}", "TXT")
+    except (dns.resolver.NoAnswer, dns.resolver.NXDOMAIN):
+        return "dns", f"no TXT record found for _acme-challenge.{host}"
+    except dns.resolver.NoNameservers as e:
+        return "dnsNoNameServers", e
+    except dns.resolver.YXDOMAIN as e:
+        return "dnsQueryTooLong", e
+    except dns.resolver.LifetimeTimeout as e:
+        return "dnsTimeout", e
+
+    # Verify the expected challenge is present
+    expect = calculate_sha256(challenge)
+    is_expected = False
+    for answer in answers:
+        if expect == str(answer).strip('"'):
+            is_expected = True
+            break
+
+    if not is_expected:
+        return "incorrectResponse", f"no token found in TXT record {expect}"
```

Comment:
```suggestion
    for answer in answers:
        if expect == answer.strings[0].decode():
            break
    else:
        return "incorrectResponse", f"no token found in TXT record {expect}"
```
(yes, that's the correct indentation for this else :) )

the `.strip('"')` makes me a bit queasy. i believe we can get to the challenge string with the code above, although we need to make sure it doesn't raise an exception (if .strings is empty or contains non-ascii characters)

Reply:
Your `else` proposal don't work if DNS return multiple TXT !? It's return error at the first answer if it don´t match.
`makes me a bit queasy`... I'm not a Python developer...
I didn't know string `decode` function, thanks.
I don't understand how works `.strings[X]`, i trust you. But managing strings with array index like this is a very bad practice, possibly a lib or language failure, idk...

Reply:
> Your else proposal don't work if DNS return multiple TXT

yes, it does. this is a for-else loop, not an if-else statement. it's a lesser used feature, but in short: if the loop is not `break`ed, the else clause of the for-loop is triggered.

`.strings` is an [attribute of the dnspython response object](https://dnspython.readthedocs.io/en/stable/rdata-subclasses.html#dns.rdtypes.ANY.TXT.TXT.strings), which is a list of the content.

Reply:
Obviously, `answers` is returned as an array of string with `resolve()` instead of `query()`